[RHCLOUD-18182] Database DAO for authentications

MikelAlejoBR opened the pull request:
This PR:

- Creates a new config option, "VaultOn", which gets set to `true` if the `VAULT_ADDR` environment variable has been set.
- Creates a database DAO to query the authentications table.
- Adapts the response depending on whether Vault is online or not. The only difference in the model is the `ID` field, which will either be filled with Vault's ID or the database's auto generated ID.

## Links

[[RHCLOUD-18182]](https://issues.redhat.com/browse/RHCLOUD-18182)

MikelAlejoBR committed:
feature: configuration option to check if Vault is online

MikelAlejoBR committed:
feature: DB DAO for authentications

MikelAlejoBR committed:
refactor: add extra field to avoid conversion errors with GORM

GORM wasn't picking up the database "extra" field correctly if the field
type was set to "map[string]interface{}". Instead, setting it to
"datatypes.JSON" flawlessly works.

MikelAlejoBR committed:
refactor: marketplace token serialization for "extraDb" element

MikelAlejoBR committed:
refactor: process "extra" field properly in auth hanlders

The "extra" field must be mapped either to "extra" or "extraDb",
depending on whether the Vault instance is online or not.

MikelAlejoBR committed:
refactor: don't stringify the marketplace token when using Vault

If the token gets serialized as a string and then assigned, the double
quotes of the resulting JSON get escaped, thus returning a very weird
JSON to the clients.

Instead, if the "BearerToken" struct is directly given, once the above
levels serialize the "extra" field properly by creating a proper JSON
object out of whatever there might be.

MikelAlejoBR committed:
refactor: use a custom "authentications" struct for creating the table

Gorm was creating FKs for the "authentications" database table, but in
none of the environments we do have such FKs. There doesn't seem to be
a way of telling Gorm to don't do that without modifying the original
model, and since that's not viable —I don't want to modify the model's
definitions because of the tests—, using an anonymous struct seems to be
the way of avoiding those FKs.

MikelAlejoBR committed:
refactor: adapt tests to the database backed authentications

Prior to these changes, the authentications were always expected to be
returned and serialized with the Vault structure, including its string
IDs.

Now, in order to make the unit tests work, we need to simulate that
Vault is online, so that the expected payloads are returned. This also
coincides to when the integration tests are run with Vault online.

For the database tests, as the ID changes, different payloads need to be
checked when running the integration tests.

MikelAlejoBR committed:
refactor: disable status listener tests for database backed auths

The tests of the status listener code fail because essentially we are
loading the JSON in memory and changing the timestamps so that we can
check if the status listener outputs the payload we expect it to output.

The problem with that approach is that in the case of database backed
authentications, things like the "Vault path" of an authentication don't
make sense, and therefore the expected structure should be different.

I initially attempted at creating a set of tests for the status listener
and the database backed authentications, but I found it very difficult
to do so. That's why I opened a PR[1] suggesting a refactor that would make
things simpler for this particular use case.

Since things are moving forward, we agreed to skip the status listener
tests and add them for the database backed authentications once the
refactor is in place.

So what I've done is simply make the status listener tests work as if we
only supported Vault, because that way they work all the time. Once the
refactor of the status listener tests is done we will be able to
properly modify the tests to include the database backed authentications
to the test mix.

[1]: https://github.com/RedHatInsights/sources-api-go/pull/139

lindgrenj6 commented on config/config.go:
```diff
@@ -105,6 +106,8 @@ func Get() *SourcesApiConfig {
 	options.SetDefault("SlowSQLThreshold", 2) //seconds
 	options.SetDefault("BypassRbac", os.Getenv("BYPASS_RBAC") == "true")
+	// We assume that if the "VAULT_ADDR" environment variable is set, Vault is online
+	options.Set("VaultOn", os.Getenv("VAULT_ADDR") != "")
```
I'm not completely in love with this - **but we can probably leave it for now, it's just the naming**.

---

But I'm wondering if instead we could have an environment variable that was `"SECRET_STORE"` (or something) and then have convenience methods on the config struct

OR just have config package level functions that look and see if the store is configured to be vault or not. e.g.

```go
func VaultOn() bool { 
	return config.SecretStore == "vault"
}
```

We'd probably default to Postgres as well - which is what we're doing here.

MikelAlejoBR replied:
Yeah, I like this idea more. I'm changing it!

lindgrenj6 commented on model/authentication.go:
```diff
@@ -153,3 +162,22 @@ func (auth *Authentication) UpdateBy(attributes map[string]interface{}) error {
 }
+
+// mapIdExtraFields returns the ID and the Extra fields ready to be assigned to the response model, depending on
+func mapIdExtraFields(auth *Authentication) (string, map[string]interface{}) {
```
Chewing on this a bit. I'm wondering if we should make a struct that just contains the "changeable depending on the secret store" fields. In this case it would just be an id + extra but maybe would expand in the future (for instance `version` wouldn't be present I dont think).

MikelAlejoBR replied:
I'm not sure I follow. What do you mean? Creating an embeddable struct that handles those fields?

lindgrenj6 replied:
I'm changing my  mind the more I think about it. I was just trying to think of a better way than just passing a tuple back and forth where we would have to know that this function basically just decides on which fields are the id/extra based on the secret store.

We can change it later if we want to.

MikelAlejoBR replied:
Yep, I don't like that function either, to be honest. Feels really ugly.

lindgrenj6 reviewed (commented):
I have a couple very minor things - but otherwise looks good can't wait to get this merged. 

I will need to test this with /bulk_create but it should "just work" after updating the ApplicationAuthentication logic in the code.

MikelAlejoBR committed:
refactor: use "SECRET_STORE" to specify if we're using the database or Vault

Instead of relying on the "VAULT_ADDR" environment variable, which makes
it harder to reason about, using a "SECRET_STORE" allows us to be more
specific and clear about what storage back end we are using for
the secrets.

lpichler reviewed (approved):
LGTM 👍
